src/module.cpp

```diff
@@ -113,6 +125,28 @@ void module::assign(const module& m)
         }
         else
         {
+            // if there are sub_module inputs, need to make a copy of the submodule
+            auto arg_modules = ins->module_inputs();
```

Comment:
The modules should only be replaced when copying the program not when copying just the module. When copying a module class it should still reference the same modules. We should add a `module_ref replace_module(module_ref mod,  module_ref rep)` method to the module class, so then the program will make a copy of all modules and then replace the module_refs with the copied modules:

```cpp
void program::assign(const program& p)
{
    if(!impl)
    {
        impl = std::make_unique<program_impl>();
    }
    else if(!impl->modules.empty())
    {
        impl->modules.clear();
    }
    impl->ctx         = p.impl->ctx;
    impl->target_name = p.impl->target_name;
    impl->modules     = p.impl->modules;
    for(auto&& mp:impl->modules) 
    {
        auto&& mod = mp.second;
        for(auto* old_mod:mod.get_sub_modules())
        {
            mod.replace_module(old_mod, &impl->modules.at(old_mod->name()));
        }
    }
}
```